sql/api/src/main/scala/org/apache/spark/sql/types/StringType.scala

```diff
@@ -75,7 +75,13 @@ class StringType private (val collationId: Int) extends AtomicType with Serializ
    */
   override def typeName: String =
     if (isUTF8BinaryCollation) "string"
-    else s"string collate ${CollationFactory.fetchCollation(collationId).collationName}"
+    else s"string collate $collationName"
+
+  override def toString: String =
+    if (isUTF8BinaryCollation) "StringType"
+    else s"StringType($collationName)"
+
+  def collationName: String = CollationFactory.fetchCollation(collationId).collationName
```

Comment:
let's be careful for adding a new public API. We can mark it as `private[sql]` first.